src/v/kafka/fetch_session.h

```diff
@@ -0,0 +1,255 @@
+        size_t operator()(const model::topic_partition& v) const {
+            return absl::Hash<model::topic_partition>{}(v);
+        }
+    };
+
+    struct topic_partition_eq {
+        using is_transparent = void;
+
+        bool operator()(
+          const model::topic_partition& lhs,
+          const model::topic_partition& rhs) const {
+            return lhs.topic == rhs.topic && lhs.partition == rhs.partition;
+        }
+
+        bool operator()(
+          const model::topic_partition_view& lhs,
+          const model::topic_partition_view& rhs) const {
+            return lhs.topic == rhs.topic && lhs.partition == rhs.partition;
+        }
+
+        bool operator()(
+          const model::topic_partition& lhs,
+          const model::topic_partition_view& rhs) const {
+            return model::topic_view(lhs.topic) == rhs.topic
+                   && lhs.partition == rhs.partition;
+        }
+        bool operator()(
+          const model::topic_partition_view& lhs,
+          const model::topic_partition& rhs) const {
+            return operator()(rhs, lhs);
+        }
+    };
+
+    using entry_ptr = ss::lw_shared_ptr<entry>;
+    using underlying_t = absl::flat_hash_map<
+      model::topic_partition_view,
+      entry_ptr,
+      topic_partition_hash,
+      topic_partition_eq>;
+
+    using io_list_t = intrusive_list<entry, &entry::_hook>;
+
+    static auto make_io_iterator(io_list_t::const_iterator it) {
+        return boost::iterators::make_transform_iterator(
+          it, [](const entry& e) { return e.partition; });
+    }
+
+public:
+    void emplace(kafka::fetch_partition v) {
+        auto e = ss::make_lw_shared<entry>(std::move(v));
+        auto [it, _] = partitions.emplace(
+          model::topic_partition_view(
+            e->partition.topic, e->partition.partition),
+          e);
+
+        insertion_order.push_back(*it->second);
+    }
+
+    bool contains(model::topic_partition_view v) {
+        return partitions.contains(v);
+    }
+
+    void erase(model::topic_partition_view v) { partitions.erase(v); }
+
+    underlying_t::iterator find(model::topic_partition_view v) {
+        return partitions.find(v);
+    }
+
+    underlying_t::const_iterator find(model::topic_partition_view v) const {
+        return partitions.find(v);
+    }
+
+    bool empty() const { return partitions.empty(); }
+
+    size_t size() const { return partitions.size(); }
+
+    auto cbegin_insertion_order() const {
+        return make_io_iterator(insertion_order.cbegin());
+    }
+
+    auto cend_insertion_order() const {
+        return make_io_iterator(insertion_order.cend());
+    }
+
+    size_t mem_usage() {
+        using debug = absl::container_internal::hashtable_debug_internal::
+          HashtableDebugAccess<underlying_t>;
+        return debug::AllocatedByteSize(partitions)
+               + partitions.size() * sizeof(fetch_partition);
+    }
+
+    underlying_t::iterator begin() { return partitions.begin(); }
```

Comment:
use type alias at the top. 

also define `begin() const ` and `end() const` 

see iobuf for alias examples.